File: python3/determine_if_two_strings_are_close.py
from collections import Counter


class Solution:
    def closeStringsCounter(self, word1: str, word2: str) -> bool:
        c1 = Counter(word1)
        c2 = Counter(word2)

        if c1.keys() != c2.keys():
            return False

        if Counter(c1.values()) != Counter(c2.values()):
            return False

        return True

    def closeStringsGeneric(self, word1: str, word2: str) -> bool:
        n = len(word1)
        m = len(word2)

        if n != m:
            return False

        d1, d2 = {}, {}
        count1, count2 = {}, {}

        for i in range(n):
            if word1[i] not in d1:
                d1[word1[i]] = 1
            else:
                d1[word1[i]] += 1

            if word2[i] not in d2:
                d2[word2[i]] = 1
            else:
                d2[word2[i]] += 1

        # len check
        if len(d1) != len(d2):
            return False

        # key check
        for k, _ in d1.items():
            if k not in d2:
                return False

        for _, v in d1.items():
            if v not in count1:
                count1[v] = 1
            else:
                count1[v] += 1

        for _, v in d2.items():
            if v not in count2:
                count2[v] = 1
            else:
                count2[v] += 1

        for key, val in count1.items():
            if key not in count2:
                return False
            if count2[key] != val:
                return False

        return True

File: golang/determine_if_two_strings_are_close.go
package main

/*
Space	: O(n)
Time	: O(n)
*/

func closeStrings(word1 string, word2 string) bool {
	h1, h2 := map[rune]int{}, map[rune]int{}
	c1, c2 := map[int]int{}, map[int]int{}

	for _, c := range word1 {
		h1[c]++
	}

	for _, c := range word2 {
		h2[c]++
	}

	for k, v := range h1 {
		if _, ok := h2[k]; !ok {
			return false
		}
		c1[v]++
	}

	for k, v := range h2 {
		if _, ok := h1[k]; !ok {
			return false
		}
		c2[v]++
	}

	for k, v := range c1 {
		if _, ok := c2[k]; !ok {
			return false
		}

		if c2[k] != v {
			return false
		}
	}

	return true
}
